thriftrw/_buffer.pyx

```diff
@@ -0,0 +1,105 @@
+# Copyright (c) 2015 Uber Technologies, Inc.
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+from __future__ import absolute_import, unicode_literals, print_function
+
+from libc.stdlib cimport malloc, realloc, free
+from libc.string cimport memcpy
+
+
+cdef int DEFAULT_CAPACITY = 4096  # 4k
+
+
+cdef class WriteBuffer(object):
```

Comment:
It seems like we should be able to do something like below, but I imagine what you have is probably equally performant. It's probably too much trouble given how much StringIO changes between 2/3. My only concern is how this implementation manages memory, because I don't know when cython calls `__dealloc__`.

```
cdef extern from "Python.h":
    ctypedef struct PyTypeObject:
        pass

cdef extern from "cStringIO.h":
    
    ctypedef struct PycStringIO_CAPI:
        PyTypeObject *InputType
        PyTypeObject *OutputType
        object (*NewInput)(object)
        object (*NewOutput)(object)
        int (*cwrite)(object, const char *, int)
        object *(*cgetvalue)(object)